### PR TITLE
Added parser and logger

### PR DESCRIPTION
**Pull request by animesh-007**
Added parser and logger. Can you review it and let me know if there are any other changes that need to be made? I also wanted to ask what are things you would like to log (like models architect, metrics) using logger?

### CODE REVIEW

**Comment by adityac8** [comment]
Model arch logging is not required.
Metrics needs logging.

File: datasets/image_util.py [hunk 1]
```diff
@@ -3,9 +3,8 @@
 import scipy.io as sio
 import torch
 from sklearn import preprocessing
-import sys
-import pdb
-import h5py
+# import sys
```
Comment (adityac8) [inline datasets/image_util.py]: Remove them

File: datasets/image_util.py [hunk 2]
```diff
@@ -44,40 +43,41 @@ def read_matdataset(self, opt):
...
-                mx = self.train_feature.max()
-                self.train_feature.mul_(1/mx)
-                self.train_label = torch.from_numpy(label[trainval_loc]).long() 
-                self.test_unseen_feature = torch.from_numpy(_test_unseen_feature).float()
-                self.test_unseen_feature.mul_(1/mx)
-                self.test_unseen_label = torch.from_numpy(label[test_unseen_loc]).long() 
-                self.test_seen_feature = torch.from_numpy(_test_seen_feature).float() 
-                self.test_seen_feature.mul_(1/mx)
-                self.test_seen_label = torch.from_numpy(label[test_seen_loc]).long()
+        # if not opt.validation:
```
Comment (adityac8) [inline datasets/image_util.py]: Seems ok, remove them.

File: datasets/image_util.py [hunk 3]
```diff
@@ -44,40 +43,41 @@ def read_matdataset(self, opt):
...
-                self.test_unseen_feature = torch.from_numpy(feature[test_unseen_loc]).float()
-                self.test_unseen_label = torch.from_numpy(label[test_unseen_loc]).long() 
-                self.test_seen_feature = torch.from_numpy(feature[test_seen_loc]).float() 
-                self.test_seen_label = torch.from_numpy(label[test_seen_loc]).long()
-        else:
-            self.train_feature = torch.from_numpy(feature[train_loc]).float()
-            self.train_label = torch.from_numpy(label[train_loc]).long()
-            self.test_unseen_feature = torch.from_numpy(feature[val_unseen_loc]).float()
-            self.test_unseen_label = torch.from_numpy(label[val_unseen_loc]).long() 
+        # else:
```
Comment (adityac8) [inline datasets/image_util.py]: These lines also removed.

File: networks/TFVAEGAN_model.py [hunk 4]
```diff
@@ -20,8 +20,9 @@ class Encoder(nn.Module):
     def __init__(self, opt):
 
         super(Encoder,self).__init__()
-        layer_sizes = opt.encoder_layer_sizes
-        latent_size = opt.latent_size
+        layer_sizes = [8192, 4096]
```
Comment (adityac8) [inline networks/TFVAEGAN_model.py]: Don't put any numbers manually.

File: networks/TFVAEGAN_model.py [hunk 5]
```diff
@@ -45,8 +46,9 @@ def __init__(self, opt):
 
         super(Generator,self).__init__()
 
-        layer_sizes = opt.decoder_layer_sizes
-        latent_size=opt.latent_size
+        layer_sizes = [4096, 8192]
```
Comment (adityac8) [inline networks/TFVAEGAN_model.py]: Don't put any numbers manually.

File: options/tfVAEGAN/CUB.yml [hunk 6]
```diff
@@ -7,23 +7,23 @@ datasets:
   name: CUB
   type: ZSLDataset
   dataroot: datasets
-
+
```
Comment (adityac8) [inline options/tfVAEGAN/CUB.yml]: remove spaces.

File: options/tfVAEGAN/CUB.yml [hunk 7]
```diff
@@ -32,15 +32,15 @@ network:
...
+    lr: 0.0001
     recons_weight: 0.01
 
   classifier:
     lr: 0.001
     gzsl: True
 
   feedback:
-    lr: 0.00001
+    lr: 0.00001
```
Comment (adityac8) [inline options/tfVAEGAN/CUB.yml]: remove spaces.

File: train_images.py [hunk 8]
```diff
@@ -10,26 +10,26 @@
...
-import math
-import sys
-from sklearn import preprocessing
-import csv
 #import functions
 import networks.TFVAEGAN_model as model
 import datasets.image_util as util
 import classifiers.classifier_images as classifier
-from config_images import opt
+from utils.logger import *
```
Comment (adityac8) [inline train_images.py]: remove * imports.

File: train_images.py [hunk 9]
```diff
@@ -10,26 +10,26 @@
...
-if opt.manualSeed is None:
-    opt.manualSeed = random.randint(1, 10000)
-print("Random Seed: ", opt.manualSeed)
-random.seed(opt.manualSeed)
-torch.manual_seed(opt.manualSeed)
-if opt.cuda:
-    torch.cuda.manual_seed_all(opt.manualSeed)
-cudnn.benchmark = True
-if torch.cuda.is_available() and not opt.cuda:
+opt = parse_options()
```
Comment (adityac8) [inline train_images.py:27]: Give option from `argparse` to pass a `yml` file.

File: train_images.py [hunk 10]
```diff
@@ -10,26 +10,26 @@
...
+
+if opt['manual_seed'] is None:
```
Comment (adityac8) [inline train_images.py]: Random seed is always passed from yml.

File: train_images.py [hunk 11]
```diff
@@ -67,14 +67,23 @@
     one = one.cuda()
     mone = mone.cuda()
 
+# making directory for logger
+os.makedirs(opt['log'],exist_ok=True)
+
+# initialize loggers
+# logger, tb_logger = init_loggers(opt)
```
Comment (adityac8) [inline train_images.py]: Why is it commented?

File: train_images.py [hunk 12]
```diff
@@ -105,71 +114,72 @@ def generate_syn_feature(generator,classes, attribute,num,netF=None,netDec=None)
...
     return syn_feature, syn_label
 
 
-optimizer          = optim.Adam(netE.parameters(), lr=opt.lr)
-optimizerD         = optim.Adam(netD.parameters(), lr=opt.lr,betas=(opt.beta1, 0.999))
-optimizerG         = optim.Adam(netG.parameters(), lr=opt.lr,betas=(opt.beta1, 0.999))
-optimizerF         = optim.Adam(netF.parameters(), lr=opt.feed_lr, betas=(opt.beta1, 0.999))
-optimizerDec       = optim.Adam(netDec.parameters(), lr=opt.dec_lr, betas=(opt.beta1, 0.999))
+optimizer          = optim.Adam(netE.parameters(), lr=opt["network"]["gan"]["lr"])
+optimizerD         = optim.Adam(netD.parameters(), lr=opt["network"]["gan"]["lr"],betas=(0.5, 0.999))
```
Comment (adityac8) [inline train_images.py]: Add betas as a list of two elements in yml file.

File: train_images.py [hunk 13]
```diff
@@ -297,10 +307,12 @@ def calc_gradient_penalty(netD,real_data, fake_data, input_att):
     netDec.train()
     netF.train()
 
-print('Dataset', opt.dataset)
+# logger.info('End of training.')
+
+print('Dataset', opt["datasets"]["name"])
```
Comment (adityac8) [inline train_images.py]: Remove print and add logger.

File: utils/logger.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+def get_time_str():
+    return time.strftime('%Y%m%d_%H%M%S', time.localtime())
+
+def init_loggers(opt):
+    log_file = os.path.join(opt['log'], f"train_{opt['name']}_{get_time_str()}.log")
+    logger = get_root_logger(logger_name='ZSl', log_level=logging.INFO, log_file=log_file)
+    # logger.info(get_env_info())
+    logger.info(dict2str(opt))
+
+    # initialize wandb logger before tensorboard logger to allow proper sync:
```
Comment (adityac8) [inline utils/logger.py]: Remove extra lines.

File: utils/logger.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    #     assert opt['logger'].get('use_tb_logger') is True, ('should turn on tensorboard when using wandb')
+    #     init_wandb_logger(opt)
+
+    tb_logger = None
+    # if opt['logger'].get('use_tb_logger'):
+    # tb_logger = init_tb_logger(log_dir=os.path.join('tensorboard_logger', opt['name']))
+    return logger, tb_logger
+
+# @master_only
+def init_tb_logger(log_dir):
```
Comment (adityac8) [inline utils/logger.py]: Remove extra lines.

File: utils/logger.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+    from torch.utils.tensorboard import SummaryWriter
+    tb_logger = SummaryWriter(log_dir=log_dir)
+    return tb_logger
+
+
+# @master_only
+def init_wandb_logger(opt):
```
Comment (adityac8) [inline utils/logger.py]: Remove extra lines.

File: utils/logger.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+
+def get_env_info():
+    """Get environment information.
+
+    Currently, only log the software version.
+    """
+    import torch
+    import torchvision
+
+    # from basicsr.version import __version__
```
Comment (adityac8) [inline utils/logger.py]: Remove extra lines.

File: utils/options.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    return msg
+
+def parse_options():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-opt', type=str, required=True, help='Path to option YAML file.')
+    args = parser.parse_args()
+    opt = parse(args.opt)
+
+    # random seed
+    seed = opt.get('manual_seed')
```
Comment (adityac8) [inline utils/options.py]: Remove extra lines.

File: utils/options.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+            msg += ' ' * (indent_level * 2) + k + ':['
+            msg += dict2str(v, indent_level + 1)
+            msg += ' ' * (indent_level * 2) + ']\n'
+        else:
+            msg += ' ' * (indent_level * 2) + k + ': ' + str(v) + '\n'
+    return msg
+
+def parse_options():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-opt', type=str, required=True, help='Path to option YAML file.')
```
Comment (adityac8) [inline utils/options.py]: Give this option in train.py

File: train_images.py [hunk 20]
```diff
@@ -10,36 +8,48 @@
...
+
+# initialize loggers
+logger= init_loggers(opt)
+
+logger.info(f"Random Seed: {opt['manual_seed']}")
+random.seed(opt['manual_seed'])
+torch.manual_seed(opt['manual_seed'])
+if torch.cuda.is_available():
+    cuda = True
+if cuda:
```
Comment (adityac8) [inline train_images.py]: remove this line.

File: networks/TFVAEGAN_model.py [hunk 21]
```diff
@@ -20,8 +19,9 @@ class Encoder(nn.Module):
     def __init__(self, opt):
 
         super(Encoder,self).__init__()
-        layer_sizes = opt.encoder_layer_sizes
-        latent_size = opt.latent_size
+        layer_sizes = opt["network"]["gan"]["layer_sizes"]
+        layer_sizes[0] = opt["network"]["gan"]["res_size"]
```
Comment (adityac8) [inline networks/TFVAEGAN_model.py]: This is confusing.

File: options/tfVAEGAN/CUB.yml [hunk 22]
```diff
@@ -7,23 +7,28 @@ datasets:
   name: CUB
   type: ZSLDataset
   dataroot: datasets
-
+
```
Comment (adityac8) [inline options/tfVAEGAN/CUB.yml]: Remove space.

File: options/tfVAEGAN/CUB.yml [hunk 23]
```diff
@@ -32,7 +37,8 @@ network:
     noise: True
 
   decoder:
-    lr: 0.001
+    layer_sizes: [4096, 8192]
+    lr: 0.0001
```
Comment (adityac8) [inline options/tfVAEGAN/CUB.yml]: Change order.

**Comment by adityac8** [comment]
See unresolved issues from before also.

File: options/tfVAEGAN/CUB.yml [hunk 24]
```diff
@@ -7,23 +7,28 @@ datasets:
...
-
+ 
 train:
   num_workers: 8
   batch_size: 64
   num_epoch: 300
 
+optimizer:
+  adam:
+    betas: [0.5, 0.999]
```
Comment (adityac8) [inline options/tfVAEGAN/CUB.yml]: Include it under `train`.
